Pontifícia Universidade Católica de São Paulo 

`Ciência de Dados e Inteligência Artificial`

🎓 Aula 2 - Processamento de Dados em Python usando Pandas

---
> 👨‍🏫*Professor Rooney Coelho (rracoelho@pucsp.br)*
---




In [1]:
# Inicialmente execute este bloco para as dependências
import pandas as pd

### 1) Dado o seguinte `DataFrame`, faça o que se pede:

In [2]:
# Aperte Ctrl+Enter
df = pd.DataFrame({'a':[11,21,31],'b':[21,22,23]})

a) Mostre as três primeiras linhas de df

In [3]:
df.head(3)

a   b
0  11  21
1  21  22
2  31  23

b) Obtenha a coluna 'a'

In [4]:
df['a']

0    11
1    21
2    31
Name: a, dtype: int64

### 2) Dado o seguinte DataFrame, faça o que se pede:

In [5]:
# Aperte Ctrl+Enter
df=pd.DataFrame({'a':[1,2,1],'b':[1,1,1]})

a) Identifique os valores únicos na coluna `a`.

In [6]:
df['a'].unique()

array([1, 2], dtype=int64)

b) Retorne um `Dataframe` com linhas somente onde a coluna `a` é menor que 2.

In [7]:
df[df['a']<2]

a  b
0  1  1
2  1  1

### 3) Carregue o Dataset do Titanic e faça o que se pede.

In [8]:
# Aperte Ctrl+Enter para carregar o Dataset
df=pd.read_csv("http://bit.ly/kaggletrain")

a) Confira os dados imprimindo o seu cabeçalho.

In [9]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

b) Confira a coluna `PassengerId` e veja se todos os seus valores são únicos.

In [10]:
df['PassengerId'].is_unique

True

b) Confira se a coluna `Ticket` possui e veja se todos os seus valores são únicos

In [11]:
df['Ticket'].is_unique

False

c) Torne a coluna `PassengerId` a chave primária de df.

In [12]:
df.set_index('PassengerId', inplace=True)

d) Utilize um método para a estatística descritiva desse `Dataset`

In [13]:
df.describe()

Survived      Pclass         Age       SibSp       Parch        Fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

### 4) Faça o que se pede.


a) Imprima os dados da pessoa mais velha que estava no navio.

In [14]:
df[df.Age == df.Age.max()]

Survived  Pclass                                  Name   Sex  \
PassengerId                                                                 
631                 1       1  Barkworth, Mr. Algernon Henry Wilson  male   

              Age  SibSp  Parch Ticket  Fare Cabin Embarked  
PassengerId                                                  
631          80.0      0      0  27042  30.0   A23        S

b) Imprima os dados do passageiro cujo número do índice é 789

In [15]:
df.loc[789]

Survived                             1
Pclass                               3
Name        Dean, Master. Bertram Vere
Sex                               male
Age                                1.0
SibSp                                1
Parch                                2
Ticket                       C.A. 2315
Fare                            20.575
Cabin                              NaN
Embarked                             S
Name: 789, dtype: object

c) Um historiador encontrou o `Ticket` 17474 do Titanic. Localize esse passageiro para entrar em contato com seus familiares e notificá-los desse achado. Essa pessoa sobreviveu?

In [16]:
df.loc[df.Ticket == '17474']

Survived  Pclass                                       Name  \
PassengerId                                                                
691                 1       1                    Dick, Mr. Albert Adrian   
782                 1       1  Dick, Mrs. Albert Adrian (Vera Gillespie)   

                Sex   Age  SibSp  Parch Ticket  Fare Cabin Embarked  
PassengerId                                                          
691            male  31.0      1      0  17474  57.0   B20        S  
782          female  17.0      1      0  17474  57.0   B20        S

**Resposta:**

*Analisando o Dataset, é possível notar que há duas pessoas que embarcaram com o ticket 17474. Contudo, essas duas pessoas contam com survived igual a 1, logo temos a certeza que o qualquer que seja o dono do ticket encontrado,* **ele sobreviveu**

### 5) Para as três classes, faça o que se pede.

a) Quantas pessoas estão na Primeira, Segunda e Terceira classes? Use a coluna `Pclass` para isso.

In [17]:
df.Pclass.value_counts()

3    491
1    216
2    184
Name: Pclass, dtype: int64

*Há 491 passageiros na terceira classe, 184 na segunda classe e 216 na primeira classe.*

b) Qual a média de idade dos sobreviventes das três classes? Use o método `groupby` para isso.

In [18]:
df[df.Survived == 1].groupby(['Pclass']).Age.mean()

Pclass
1    35.368197
2    25.901566
3    20.646118
Name: Age, dtype: float64

c) Qual a média de idade das pessoas que sobreviveram e morreram nas três classes? Use o método `groupby` para isso (note que são dois agrupamentos).

In [19]:
df.groupby(['Survived', 'Pclass']).Age.mean()

Survived  Pclass
0         1         43.695312
          2         33.544444
          3         26.555556
1         1         35.368197
          2         25.901566
          3         20.646118
Name: Age, dtype: float64

d) Qual a média de idade das pessoas de cada sexo que sobreviveram e morreram nas três classes? Use o método `groupby` para isso (note que são três agrupamentos).

In [20]:
df.groupby(['Survived', 'Pclass', 'Sex']).Age.mean()

Survived  Pclass  Sex   
0         1       female    25.666667
                  male      44.581967
          2       female    36.000000
                  male      33.369048
          3       female    23.818182
                  male      27.255814
1         1       female    34.939024
                  male      36.248000
          2       female    28.080882
                  male      16.022000
          3       female    19.329787
                  male      22.274211
Name: Age, dtype: float64

### 6) Use o método `replace` para mudar a coluna `Survived` pela correspondência do seguinte dicionário:

```python
di = {0: False, 1: True}
```

https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.replace.html

In [21]:
di = {0: False, 1: True}

In [22]:
df['Survived'] = df['Survived'].replace(di)

In [23]:
df

Survived  Pclass  \
PassengerId                     
1               False       3   
2                True       1   
3                True       3   
4                True       1   
5               False       3   
...               ...     ...   
887             False       2   
888              True       1   
889             False       3   
890              True       1   
891             False       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   
...                                                        ...     ...   ...   
887                                      Montvila, Rev. Juozas    male  27.0   
888                               Graham, Miss. Margaret Edith  female  19.0   
889                   Johnston, Miss. Catherine Helen "Carrie"  female   NaN   
890                                      Behr, Mr. Karl Howell    male  26.0   
891                                        Dooley, Mr. Patrick    male  32.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   NaN        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
4                1      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   NaN        S  
...            ...    ...               ...      ...   ...      ...  
887              0      0            211536  13.0000   NaN        S  
888              0      0            112053  30.0000   B42        S  
889              1      2        W./C. 6607  23.4500   NaN        S  
890              0      0            111369  30.0000  C148        C  
891              0      0            370376   7.7500   NaN        Q  

[891 rows x 11 columns]